Брать чаевые и общий счёт из GET-параметров

skrattaren opened the pull request.

skrattaren committed:
Брать чаевые и общий счёт из GET-параметров

ElenaZimina commented on script.js:
```diff
@@ -1,3 +1,20 @@
+var prmstr = window.location.search.substr(1);
```
`var` используется в стандарте языка ES5. Поскольку скрипт написан с использованием стандарта ES6+, лучше заменить `var` на `const`.

ElenaZimina replied:
Вообще, чтобы не наполнять глобальную область видимости новыми переменными (которые, к тому же, становятся свойствами объекта `window` 👻), что считается плохой практикой, стоит поместить весь новый код внутрь самовызываемой функции, которая начинается со строки 18. 
Смысл существования этой функции в том, чтобы все созданные в скрипте переменные остались локальными, а не стали глобальными.

ElenaZimina commented on script.js:
```diff
@@ -1,3 +1,20 @@
+var prmstr = window.location.search.substr(1);
+var params = {};
+var prmarr = prmstr.split("&");
+for ( var i = 0; i < prmarr.length; i++) {
```
В данной строке переменная `i` меняется, поэтому лучше заменить декларацию `var` на `let`. Также удали, пожалуйста, пробел после `(`.

ElenaZimina commented on script.js:
```diff
@@ -1,3 +1,20 @@
+var prmarr = prmstr.split("&");
+for ( var i = 0; i < prmarr.length; i++) {
+	var tmparr = prmarr[i].split("=");
+	if ( ! isNaN(tmparr[1]) ) {
```
Лишние пробелы вокруг `!`.

ElenaZimina commented on script.js:
```diff
@@ -1,3 +1,20 @@
+	}
+}
+
+if ( params['total'] != null ) {
```
Лучше сравнивать с `undefined`. 
В спецификации это, конечно, не определено, но большинством разработчиков принято, что `undefined` - это что-то неизвестное, неопределенное. А `null` - то, что разработчик обнуляет намеренно.

skrattaren committed:
Афтер ревью фиксез!

oshibka404 commented on script.js:
```diff
@@ -1,4 +1,21 @@
+		}
+	}
+
+	if (params['total'] != undefined) {
```
Для доступа к полям, имена которых известны, принято использовать dot notation вместо bracket notation. А `[]` — для случаев, когда имя поля хранится где-то в переменной.

По возможности всегда следует использовать `===` вместо `==` при сравнении, чтобы не дать приведению типов себя запутать.

И при проверке на `undefined` надёжнее использовать `typeof`, потому что ТЕОРЕТИЧЕСКИ где-то выше в коде может встретиться вполне валидная строчка `undefined = 'defined'`, хотя на практике вряд ли такое случится, конечно.

Короче, вот так:
```suggestion
	if (typeof params.total !== 'undefined') {
```

skrattaren replied:
> Для доступа к полям, имена которых известны, принято использовать dot notation вместо bracket notation. А [] — для случаев, когда имя поля хранится где-то в переменной.

Я просто сначала не терял надежды сделать это циклом, но всё-таки у полей разные названия (-:Е

oshibka404 commented on script.js:
```diff
@@ -1,4 +1,21 @@
+	if (params['total'] != undefined) {
+		document.getElementById('people__total').innerText = params['total'];
+	}
+	if (params['tips'] != undefined) {
```
то же самое здесь:
```suggestion
	if (typeof params.tips !== 'undefined') {
```

oshibka404 commented on script.js:
```diff
@@ -1,4 +1,21 @@
+	const params = {};
+	const prmarr = prmstr.split("&");
+	for (let i = 0; i < prmarr.length; i++) {
+		var tmparr = prmarr[i].split("=");
```
- `var` вообще [не стоит использовать](https://blog.usejournal.com/awesome-javascript-no-more-var-working-title-999428999994) в современном JS. Лучше использовать `let` если значение переменной меняется и `const` если не меняется
- Тут было бы удобно использовать [деструктурирующее присваивание](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment) — а то `params[tmparr[0]] = tmparr[1]` выглядит непонятненько.
```suggestion
		const [paramName, paramValue] = prmarr[i].split("=");
```

oshibka404 commented on script.js:
```diff
@@ -1,4 +1,21 @@
 (function() {
+	const prmstr = window.location.search.substr(1);
```
замечание по неймингу:
- в JS (вернее, во всех сколько-нибудь популярных стайлгайдах по JS) принято использовать `camelCase` в именах переменных.
- `prm` можно развернуть до полноценного `param`, a `str` — до `String`, а то мне понадобилась пара секунд чтобы это расшифровать имя (param string? primestar? poor master?).
```suggestion
	const paramString = window.location.search.substr(1);
```

skrattaren committed:
В стиле Андрея!

- обращение к полям объекта через точку
- больше знаков равно!
- всем `typeof`, посоны

Co-Authored-By: Andrey Ozornin <andrey.ozornin@genestack.com>

skrattaren committed:
Поименовать новые переменные попонятнее

oshibka404 commented on script.js:
```diff
@@ -1,4 +1,21 @@
+	}
+
+	if (typeof params.total !== 'undefined') {
+		document.getElementById('people__total').innerText = params['total'];
```
ой, здесь не заметил
```suggestion
		document.getElementById('people__total').innerText = params.total;
```

oshibka404 commented on script.js:
```diff
@@ -1,4 +1,21 @@
+		document.getElementById('people__total').innerText = params['total'];
+	}
+	if (typeof params.tips !== 'undefined') {
+		document.getElementById('tips').innerText = params['tips'];
```
```suggestion
		document.getElementById('tips').innerText = params.tips;
```
Когда пишешь через точку, то и подсветка синтаксиса лучше выглядит

oshibka404 commented on script.js:
```diff
@@ -1,4 +1,21 @@
 (function() {
+	const paramString = window.location.search.substr(1);
+	const params = {};
+	const paramArray = paramString.split("&");
```
ну и пусть останется только один вид кавычек
```suggestion
	const paramArray = paramString.split('&');
```

oshibka404 commented on script.js:
```diff
@@ -1,4 +1,21 @@
+	const params = {};
+	const paramArray = paramString.split("&");
+	for (let i = 0; i < paramArray.length; i++) {
+		const [paramName, paramValue] = paramArray[i].split("=");
```
```suggestion
		const [paramName, paramValue] = paramArray[i].split('=');
```

oshibka404 commented on script.js:
```diff
@@ -1,4 +1,21 @@
 (function() {
+	const paramString = window.location.search.substr(1);
```
```suggestion
	const paramString = window.location.search.substr(1); // cut off `?` symbol
```
а то тоже не самая очевидная строчка

oshibka404 reviewed (changes requested):
Ой, сорян, я тогда написал комментов и не засабмитил

skrattaren committed:
Прокомментировать неочевидную строчку

Co-Authored-By: Andrey Ozornin <andrey.ozornin@genestack.com>

skrattaren committed:
Прокомментировать неочевидную строчку

Co-Authored-By: Andrey Ozornin <andrey.ozornin@genestack.com>

skrattaren committed:
Использовать только одинарные кавычки

Co-Authored-By: Andrey Ozornin <andrey.ozornin@genestack.com>

skrattaren committed:
Только точки, только хардкор!

Co-Authored-By: Andrey Ozornin <andrey.ozornin@genestack.com>
